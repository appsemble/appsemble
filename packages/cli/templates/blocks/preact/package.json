{
  "private": true,
  "type": "module",
  "dependencies": {
    "@appsemble/preact": "0.23.4",
    "@appsemble/sdk": "0.23.4",
    "preact": "^10.0.0"
  }
}
