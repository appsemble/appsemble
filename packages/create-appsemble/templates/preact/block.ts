declare module '@appsemble/sdk' {
  interface EventListeners {
    /**
     * This event can be used to receive incoming data to display.
     */
    data: never;
  }

  interface Parameters {
    /**
     * A list of fields to render out in a table.
     */
    fields: string[];
  }
}

export {};
