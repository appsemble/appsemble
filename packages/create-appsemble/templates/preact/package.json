{
  "private": true,
  "dependencies": {
    "@appsemble/preact": "0.19.9",
    "@appsemble/sdk": "0.19.9",
    "preact": "^10.0.0"
  }
}
