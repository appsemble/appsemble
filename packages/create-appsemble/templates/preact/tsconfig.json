{
  "extends": "../../tsconfig",
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "preact"
  },
  "lib": ["dom", "dom.iterable", "esnext"],
  "types": ["@appsemble/webpack-config/types"]
}
