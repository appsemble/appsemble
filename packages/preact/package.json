{
  "name": "@appsemble/preact",
  "version": "0.19.9",
  "description": "Build your own blocks using Preact",
  "keywords": [
    "app",
    "apps",
    "appsemble",
    "framework",
    "front-end",
    "frontend",
    "low-code",
    "lowcode",
    "preact",
    "sdk"
  ],
  "homepage": "https://appsemble.app",
  "bugs": "https://gitlab.com/appsemble/appsemble/issues",
  "repository": {
    "type": "git",
    "url": "https://gitlab.com/appsemble/appsemble.git",
    "directory": "packages/preact"
  },
  "license": "LGPL-3.0-only",
  "author": "Appsemble <info@appsemble.com> (https://appsemble.com)",
  "sideEffects": false,
  "main": "dist",
  "files": [
    "dist"
  ],
  "scripts": {
    "prepack": "tsc -p tsconfig.build.json",
    "test": "jest"
  },
  "peerDependencies": {
    "@appsemble/sdk": "0.19.9",
    "preact": "^10"
  }
}
