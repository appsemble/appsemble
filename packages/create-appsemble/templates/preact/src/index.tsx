import { bootstrap } from '@appsemble/preact';
import { useState } from 'preact/hooks';

bootstrap(({ events, parameters: { fields } }) => {
  const [data, setData] = useState<any[]>(null);
  const [error, setError] = useState(false);

  useEffect(() => {
    const onData = (newData: unknown[], newError: unknown): void => {
      if (newError) {
        setError(true);
        setData(null);
      } else {
        setError(false);
        setData(newData);
      }
    };

    events.on.data(onData);

    return () => {
      events.off.data(onData);
    };
  }, [events]);

  if (error) {
    return <p>Error loading data.</p>;
  }

  if (!data) {
    return <p>Loadingâ€¦</p>;
  }

  if (!data.length) {
    return <p>No data to display.</p>;
  }

  return (
    <table className="table">
      <thead>
        <tr>
          {fields.map((field) => (
            <th key={field}>{field}</th>
          ))}
        </tr>
      </thead>
      <tbody>
        {data.map((item, dataIndex) => (
          <tr key={item.id || dataIndex}>
            {fields.map((field) => (
              <td key={field}>{item[field]}</td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
});
