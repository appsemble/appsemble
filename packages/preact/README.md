# ![](https://gitlab.com/appsemble/appsemble/-/raw/0.23.4/config/assets/logo.svg) Appsemble Preact SDK

> Build your own blocks using Preact

[![npm](https://img.shields.io/npm/v/@appsemble/preact)](https://www.npmjs.com/package/@appsemble/preact)
[![GitLab CI](https://gitlab.com/appsemble/appsemble/badges/0.23.4/pipeline.svg)](https://gitlab.com/appsemble/appsemble/-/releases/0.23.4)
[![Prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://prettier.io)

## Table of Contents

- [Installation](#installation)
- [Usage](#usage)
  - [`bootstrap()`](#bootstrap)
  - [`useBlock()`](#useblock)
  - [`<FormattedMessage />`](#formattedmessage-)
- [License](#license)

## Installation

```sh
npm install @appsemble/preact preact
```

## Usage

This package integrates [Preact](https://preactjs.com) with
[`@appsemble/sdk`](https://www.npmjs.com/package/@appsemble/sdk). Please read the documentation for
a better understanding of how to use it.

### `bootstrap()`

The bootstrap function takes a Preact component as an argument and renders. It takes the Appsemble
block context as its props, extended with the `ready()` function. The `ready()` function needs to be
called when the block is done rendering. Actions won’t be finalized for any blocks on the page,
until all blocks are ready.

```tsx
import { bootstrap } from '@appsemble/preact';
import { useEffect } from 'preact/hooks';

bootstrap(({ actions }) => {
  useEffect(() => {
    ready();
  }, []);

  return (
    <button onClick={() => actions.onClick()} type="button">
      Hello world!
    </button>
  );
});
```

**Note**: The script will be loaded only once. The component is then bootstrapped for every instance
of your block type that is loaded by the app creator.

### `useBlock()`

Get the block context. This is available for components that are rendered within a subtree of a
component rendered by [`bootstrap`](#bootstrap)

```tsx
import { useBlock } from '@appsemble/preact';
import { type VNode } from 'preact';

export function MyButton(): VNode {
  const { actions } = useBlock();

  return (
    <button onClick={() => actions.onClick()} type="button">
      Hello world!
    </button>
  );
}
```

### `<FormattedMessage />`

This helper component renders a translated message ID.

```tsx
import { FormattedMessage } from '@appsemble/preact';
import { type VNode } from 'preact';

interface MyButtonProps {
  readonly name: string;
}

export function MyButton({ name }: MyButtonProps): VNode {
  return (
    <button type="button">
      <FormattedMessage id="myButtonLabel" values={{ name }} />
    </button>
  );
}
```

## License

[LGPL-3.0-only](https://gitlab.com/appsemble/appsemble/-/blob/0.23.4/LICENSE.md) ©
[Appsemble](https://appsemble.com)
