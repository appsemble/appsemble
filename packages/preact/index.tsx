import { type BootstrapParams, type Messages, bootstrap as sdkBootstrap } from '@appsemble/sdk';
import { type ComponentChild, type ComponentType, createContext, render, type VNode } from 'preact';
import { useContext } from 'preact/hooks';

export interface BlockProps extends BootstrapParams {
  /**
   * A function that must be called to indicate the block is ready to be rendered.
   */
  ready: () => void;
}

// eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error
// @ts-ignore 2345 argument of type is not assignable to parameter of type (strictNullChecks)
const Context = createContext<BlockProps>(null);

/**
 * Mount a Preact component returned by a bootstrap function in the shadow DOM of a block.
 *
 * @param Component The Preact component to mount.
 * @returns A promise which gets resolved if the component calls `ready()`.
 */
export function mount(
  Component: ComponentType<BlockProps>,
): (params: BootstrapParams) => Promise<void> {
  return (params) =>
    new Promise((ready) => {
      const props = {
        ...params,
        ready,
      };
      const component = (
        <Context.Provider value={props}>
          <Component {...props} />
        </Context.Provider>
      );
      render(component, params.shadowRoot);
      params.utils.addCleanup(() => render(null, params.shadowRoot));
    });
}

export function bootstrap(Component: ComponentType<BlockProps>): void {
  sdkBootstrap(mount(Component));
}

/**
 * A HOC which passes the Appsemble block values to he wrapped Preact component.
 *
 * @param Component The Preact component to wrap.
 * @returns The wrapper component.
 * @deprecated Use `useBlock()` instead.
 */
export function withBlock<P extends {}>(
  Component: ComponentType<Omit<BlockProps, keyof P> & P>,
): ComponentType<P> {
  return (props: P) => (
    <Context.Consumer>{(values) => <Component {...props} {...values} />}</Context.Consumer>
  );
}

export function useBlock(): BlockProps {
  return useContext(Context);
}

interface FormattedMessagePropsWithoutValues<M extends keyof Messages> {
  /**
   * The ID of the message to render.
   */
  id: M;
}

interface FormattedMessagePropsWithValues<M extends keyof Messages> {
  /**
   * The ID of the message to render.
   */
  id: M;

  /**
   * Values to pass to the formatted message.
   */
  values: Messages[M] extends never ? undefined : Messages[M];
}

/**
 * Render an Appsemble message.
 */
export function FormattedMessage<M extends keyof Messages>(
  props: Messages[M] extends never
    ? FormattedMessagePropsWithoutValues<M>
    : FormattedMessagePropsWithValues<M>,
): VNode {
  const { utils } = useBlock();

  // @ts-expect-error The messages interface isnâ€™t implemented within the Preact SDK.
  return utils.formatMessage(props.id, props.values) as ComponentChild as VNode;
}
