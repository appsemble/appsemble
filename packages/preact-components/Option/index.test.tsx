import { render, screen } from '@testing-library/preact';
import { expect, it } from 'vitest';

import { Option } from './index.js';

it('should render an option', () => {
  render(<Option data-testid="option" value="test" />);
  const option = screen.getByTestId('option');
  expect(option).toMatchSnapshot();
});

it('should render a hidden option', () => {
  render(<Option data-testid="option" hidden value="test" />);
  const option = screen.getByTestId('option');
  expect(option.classList).toContain('is-hidden');
});
