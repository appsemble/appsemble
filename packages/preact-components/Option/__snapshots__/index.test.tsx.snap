// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render an option 1`] = `
<option
  data-testid="option"
  value="test"
/>
`;
